cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+	// when the rollout status is progressing or failed and the status remains for longer than the timeout
+	// then the status will be set to timeout
+	TimeOut
+	// skip rollout on this cluster
+	Skip RolloutStatus = iota
```

Comment:
Not sure, what is the case for skip ?

Reply:
Add this because in the addon upgrade, when the user defines 2 placements, there might be clusters that overlap. Since each cluster should only belong to one placement, it should be skipped when going through another placement.